xmr-btc/src/alice.rs

```diff
@@ -146,15 +150,30 @@ where
                 scalar: s_a.into_ed25519(),
             });
 
+            let public_spend_key = S_a + S_b_monero;
+            let public_view_key = v.public();
+
             co.yield_(Action::LockXmr {
                 amount: xmr,
-                public_spend_key: S_a + S_b_monero,
-                public_view_key: v.public(),
+                public_spend_key,
+                public_view_key,
             })
             .await;
 
-            // TODO: Watch for LockXmr using watch-only wallet. Doing so will prevent Alice
-            // from cancelling/refunding unnecessarily.
+            let monero_joint_address = monero::Address::standard(
+                monero::Network::Mainnet,
+                public_spend_key,
+                public_view_key.into(),
+            );
+
+            if let Either::Right(_) = select(
+                monero_client.watch_for_transfer_improved(monero_joint_address, xmr, v),
```

Comment:
Shouldn't Bob be the one watching for the transfer?
Before: Alice locks xmr, sends a transfer proof to Bob, Bob assumes xmr is locked because of the proff
After: Alice locks xmr, Bob watches the blockchain using the view key to confirm the xmr is locked.

Reply:
You are correct that Bob has to watch the blockchain using the view key.

But in this generator-based approach, for Alice we yield an action which may or may not be performed by the consumer of this generator, so we need to verify that this happened before we continue. Now, this is not strictly necessary because one can assume that if we receive the redeem encrypted signature from Bob then the monero must have been locked up correctly.

Therefore, in your re-implementation of the protocol I think you can safely omit this if you want.